File: 2692-take-gifts-from-the-richest-pile/2692-take-gifts-from-the-richest-pile.cpp
class Solution {
public:
    long long pickGifts(vector<int>& gifts, int k) {
        priority_queue<int> pq;

        for(auto &ele: gifts)
        {
            pq.push(ele);
        }

        while(!pq.empty() && k--)
        {
            int maxGifts = pq.top();
            if(maxGifts == 1)
                break;
                
            pq.pop();


            int leave = sqrt(maxGifts);
            pq.push(leave);
        }

        long long ans = 0;
        while(!pq.empty())
        {
            ans += pq.top();
            pq.pop();
        }

        return ans;
    }
};

File: 2692-take-gifts-from-the-richest-pile/2692-take-gifts-from-the-richest-pile.cs
public class Solution {
    public long PickGifts(int[] gifts, int k) {
        PriorityQueue<int, int> pq = new PriorityQueue<int, int>(Comparer<int>.Create((a, b) => b - a));
        
        foreach (int gift in gifts) {
            pq.Enqueue(gift, gift);
        }
        
        while (k-- > 0 && pq.Count > 0) {
            int maxGifts = pq.Dequeue();
            if (maxGifts == 1) {
                pq.Enqueue(1, 1); // Re-add 1 back to the queue
                break;
            }
            
            int leave = (int)Math.Sqrt(maxGifts);
            pq.Enqueue(leave, leave);
        }
        
        long ans = 0;
        while (pq.Count > 0) {
            ans += pq.Dequeue();
        }
        
        return ans;
    }
}
